driver/src/main/java/org/neo4j/driver/internal/messaging/PackStreamMessageFormatV1.java

```diff
@@ -484,76 +522,111 @@ private Value unpackValue() throws IOException
+            String startUrn = unpacker.unpackString();
+            String endUrn = unpacker.unpackString();
+            String relType = unpacker.unpackString();
+            Map<String,Value> props = unpackMap();
+
+            return new RelationshipValue( new SimpleRelationship( urn, startUrn, endUrn, relType, props ) );
+        }
+
+        private SimpleNode unpackNode() throws IOException
+        {
+            String urn = unpacker.unpackString();
+
+            int numLabels = (int) unpacker.unpackListHeader();
+            List<String> labels = new ArrayList<>( numLabels );
+            for ( int i = 0; i < numLabels; i++ )
+            {
+                labels.add( unpacker.unpackString() );
+            }
+            int numProps = (int) unpacker.unpackMapHeader();
+            Map<String,Value> props = new HashMap<>();
+            for ( int j = 0; j < numProps; j++ )
+            {
+                String key = unpacker.unpackString();
+                props.put( key, unpackValue() );
+            }
+
+            return new SimpleNode( urn, labels, props );
+        }
+
+        private Value unpackPath() throws IOException
+        {
+            // List of unique nodes
+            Node[] uniqNodes = new Node[(int) unpacker.unpackListHeader()];
+            for ( int i = 0; i < uniqNodes.length; i++ )
+            {
+                ensureCorrectStructSize( "NODE", NODE_FIELDS, unpacker.unpackStructHeader() );
+                ensureCorrectStructSignature( "NODE", NODE, unpacker.unpackStructSignature() );
+                uniqNodes[i] = unpackNode();
+            }
+
+            // List of unique relationships, without start/end information
+            SimpleRelationship[] uniqRels = new SimpleRelationship[(int) unpacker.unpackListHeader()];
+            for ( int i = 0; i < uniqRels.length; i++ )
+            {
+                ensureCorrectStructSize( "RELATIONSHIP", 3, unpacker.unpackStructHeader() );
+                ensureCorrectStructSignature( "UNBOUND_RELATIONSHIP", UNBOUND_RELATIONSHIP, unpacker.unpackStructSignature() );
+                Identity urn = Identities.identity( unpacker.unpackString() );
+                String relType = unpacker.unpackString();
+                Map<String,Value> props = unpackMap();
+                uniqRels[i] = new SimpleRelationship( urn, null, null, relType, props );
+            }
+
+            // Path sequence
+            int length = (int) unpacker.unpackListHeader();
+
+            // Knowing the sequence length, we can create the arrays that will represent the nodes, rels and segments in their "path order"
+            Path.Segment[] segments = new Path.Segment[length / 2];
+            Node[] nodes = new Node[segments.length + 1];
+            Relationship[] rels = new Relationship[segments.length];
+
+            Node prevNode = uniqNodes[0], nextNode; // Start node is always 0, and isn't encoded in the sequence
+            nodes[0] = prevNode;
```

Comment:
Proposal to the fix for test failure MessageFormatTest.shouldGiveHelpfulErrorOnMalformedNodeStruct. Though I do not totally understand the math to calculate the size of array here.